### PR TITLE
Enable nullable on annotations

### PR DESCRIPTION
**Pull request by tlakollo**
Enable nullable on Annotations.cs
Change the MarkStep condition to check for null after TryResolve

Fixes https://github.com/dotnet/linker/issues/2274
I couldn't add a test case because I couldn't come up with a scenario in which the TypeReference Type.BaseType is not null but resolving the TypeReference is null

### CODE REVIEW

File: src/linker/Linker/Annotations.cs [hunk 1]
```diff
@@ -244,7 +245,7 @@ public bool IsInstantiated (TypeDefinition type)
 			return marked_instantiated.Contains (type);
 		}
 
-		public void MarkRelevantToVariantCasting (TypeDefinition type)
+		public void MarkRelevantToVariantCasting (TypeDefinition? type)
```
Comment (sbomer) [inline src/linker/Linker/Annotations.cs:248]: I'd prefer to make the type non-nullable, and require callers to decide what to do if for example a type doesn't resolve.

**Review by sbomer — APPROVED** [review]
Thanks, LGTM! I think we will need to make a separate change in the 6.0 branch that's a more targeted fix for the BaseType issue (since none of the nullable annotation changes exist in that branch).

File: src/linker/Linker.Steps/MarkStep.cs [hunk 2]
```diff
@@ -1817,9 +1817,9 @@ protected internal virtual TypeDefinition MarkType (TypeReference reference, Dep
 			MarkCustomAttributes (type, new DependencyInfo (DependencyKind.CustomAttribute, type));
 			MarkSecurityDeclarations (type, new DependencyInfo (DependencyKind.CustomAttribute, type));
 
-			if (type.BaseType is not null &&
+			if (_context.TryResolve (type.BaseType) is TypeDefinition baseType &&
```
Comment (agocke) [inline src/linker/Linker.Steps/MarkStep.cs:1820]: Is nullable actually enabled in this file?

Comment (tlakollo) [reply src/linker/Linker.Steps/MarkStep.cs:1820]: Nullable is not enabled on MarkStep, but methods in Annotations.cs that have non-nullable parameters are flagged when called with a possible null. The warning is on callsite so the warning appears in MarkStep even if nullable has not been enabled.

**Pull request merged**